### DISCUSSION TITLE
Various core modifications do not work before downloading Symbols

### DISCUSSION THREAD

**Discussion by antoniu200** [valinet/ExplorerPatcher#1274] [General]
Hello,

I have tried multiple versions of Windows 11 22H2, from 22621.525 to 22621.1485 inclusive. When selecting the Windows 10 style taskbar, the Win-X menu does not work before any Symbols are downloaded on any of the versions I tested: 22621.525, 22621.1105, 22621.1413 and 22621.1485. For this test, I used the latest version of ExplorerPatcher currently available: 22621.1413.54.5.

On Windows 11 22621.525, the normal start menu also does not open before downloading Symbols, but works flawlessly after. The alternative fix is deleting dxgi.dll from StartMenuExperienceHost's install folder.

Can't these Symbols be included with the installer, eliminating the need for a separate download?

EDIT: Also take a look at #1211, #1230, #1267.

_P.S.: Where could I look into what I would need to know to be able to contribute to this project in the form of solutions to issues and maybe even implementing the solutions in the source code?_

**Comment by valinet** [comment]
> Can't these Symbols be included with the installer, eliminating the need for a separate download?

https://github.com/valinet/ExplorerPatcher/wiki/Symbols

> P.S.: Where could I look into what I would need to know to be able to contribute to this project in the form of solutions to issues and maybe even implementing the solutions in the source code?

?

**Reply by antoniu200** [reply]
> https://github.com/valinet/ExplorerPatcher/wiki/Symbols

Okay, that explains a lot, but not why issues specific to a lack of Symbols appear on releases of Windows older than the one used to test ExplorerPatcher (situation above: EP 22621.1413.54.5 on Windows 22621.525). According to the link, supported releases should already have Symbols right in the installer. I will try and get you some logs from EP on Windows 11 22621.525, those should help.

> ?

Where could I learn the things required to contribute to this project? How can ExplorerPatcher tell explorer.exe (and other programs) what functions to use for certain things? How would I know which functions are the ones I want? Basically, this https://github.com/valinet/ExplorerPatcher/wiki/How-does-it-work.

**Reply by valinet** [reply]
The document says what should usually be the case. The source code tells the actual story. I never understood why people don’t check that first, so as not to wait for a reply that might never come.

Intuitively, let’s check the ‘symbols.c’ source code file.
https://github.com/valinet/ExplorerPatcher/blob/master/ExplorerPatcher/symbols.c#L780
Around there there’s a list of file hashes, and for each some hardcoded addresses. Those are the built-in addresses, for files which have that hash (besides them is a comment about which build they are found on). If 525’s files do not match any of those hashes, then it does not contain symbols. Furthermore, no one tests on builds this old - maybe in the mean time a bug slipped in where some things do not work conditioned by something that’s available only when symbol data is present. Idk. The source code is your best friend, that’s why it’s there. Setup a development environment on your machine and debug the program and you’ll see what happens, I personally do not have a guess and it’s hard to guess considering what’s in there.

> Where could I learn the things required to contribute to this project? How can ExplorerPatcher tell explorer.exe (and other programs) what functions to use for certain things? How would I know which functions are the ones I want?

I mostly learnt pointers about the concepts needed during school (university), and the fact that those things actually exist and what they are (and for a subset of them in-depth how they work), then expanded the knowledge via free resources online (things I find via Google), and also specifics about explorer’s architecture from Microsoft’s Windows documentation (this gives mostly the big picture), the disassembly of system files (this gives the best in-depth picture), the disassembly and/or source code of other programs that do similar things to what ExplorerPatcher does (7+TT/Windhawk mods, StartAllBack etc), and finally, again, various articles around the Internet that describe certain hacks and tweaks.

I am not a teacher and everyone's mileage may vary. What I personally recommend people (and this also comes from the fact that I enjoy learning and knowing about computers, how they work etc) is to get a solid understanding on how computers actually work (including what happens to source code after it’s written by the programmer). ”Operating system”, “kernel”, “memory management unit”, “function/method”, “this”, “stack”, “heap”, “memory layout”, “shellcode”, “structs”, “pointers”, “address space”, “program land”, “kernel land”, “static”, “runtime”, “ahead of time”, “compile time”, “just in
time” are a few of the keywords that should instantly blink a light. Starting points:

https://ocw.cs.pub.ro/courses/programare
https://ocw.cs.pub.ro/courses/sd-ca
https://web.archive.org/web/20170122040821/https://ocw.cs.pub.ro/courses/iocla
https://ocw.cs.pub.ro/courses/so
https://ocw.cs.pub.ro/courses/cpl

I strongly believe formal education helps establishing a common ground, but it’s not the only way, indeed. Thorough understanding of those basic building blocks unlock the understanding of what ExplorerPatcher is: a collection of patches to explorer’s memory at a proper depth level so as to start exploring the existing source code and decipher what these ‘patches’ represent, how they are done, how do they work etc. Then, to develop new ones, as I said, you have to study  Microsoft’s implementations, which are most of the times source-less, so operating a tool like IDA or Ghidra, where you disassemble and decompile parts of the system files, and then try to understand it, also helped by the symbol files is a must. Ofc, to understand how things there work, it’s also necessary to have at least a basic idea about how Windows works and about its design, a basic idea on which to expand with more knowledge from articles found on the Internet, or, as usual, fine books that walk you through in-depth, like “Windows Internals”.

I could throw in a bunch of more in-depth jargon here, but without knowledge about the technical concepts, it’s meaningless. Most of the stuff in ExplorerPatcher is basic from an implementation point of view once you understand what concepts it builds on. There’s also an added layer of knowledge about how/what to patch which is developed, as I said, from here and there, Microsoft's docs, disassembly, various articles etc and that’s it mostly. Lastly, there’s a layer of auxiliary stuff that does various things, like the Properties UI that’s useless from the program’s main point of view, but it’s there to help make applying settings easier and so on.

Kind of difficult to give a one-line exact pointer as to where to go and find exactly what you want to do.
